### ISSUE TITLE
Unexpected keyword argument for Generic dataclass with ABC bounded TypeVar

### ISSUE THREAD

**Issue by jtblp** [pylint-dev/pylint#9488]
### Bug description

```python
The following minimal example causes pylint to fail (mypy is fine with it):


from abc import ABC
from dataclasses import dataclass
from typing import Generic, TypeVar


class Base:
    pass


class AbstractBase(ABC):
    pass


_T = TypeVar("_T", bound=Base | AbstractBase)


@dataclass
class GenericClass(Generic[_T]):
    field: _T


@dataclass
class ConcreteGeneric(GenericClass[Base | AbstractBase]):
    pass


a = ConcreteGeneric(field=Base())
```
```


### Configuration

_No response_

### Command used

```shell
pylint pylint_test.py
```

### Pylint output

```python
************* Module pylint_test
pylint_test.py:27:4: E1123: Unexpected keyword argument 'field' in constructor call (unexpected-keyword-arg)

-----------------------------------
Your code has been rated at 6.15/10
```

### Expected behavior

I would expect that this code does not raise a warning as the parameter exists.

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.3
Python 3.12.1 (main, Feb  6 2024, 10:17:43) [Clang 15.0.0 (clang-1500.1.0.2.5)]
```

### OS / Environment

MacOS

**Comment by RomanValov** [comment]
I have bumped into similar issues with dataclass inheritance with generic. Here is a minimal example to reproduce the issue:

```
from dataclasses import dataclass

@dataclass
class DataBase[T]:
    x: int
    y: int

@dataclass
class Data(DataBase[str]):
    pass

if __name__ == '__main__':
    data = Data(0, 0)
    print(data)
    data = Data(x=0, y=0)
    print(data)
```

The snippet is perfectly valid python code. But pylint ends up with errors:

```
data.py::11: E1121: Too many positional arguments for constructor call (too-many-function-args)
data.py::11: E1123: Unexpected keyword argument 'x' in constructor call (unexpected-keyword-arg)
data.py::11: E1123: Unexpected keyword argument 'y' in constructor call (unexpected-keyword-arg)
```

**Comment by mharding-hpe** [comment]
I have also seen similar-ish issues, even with the currently latest pylint.

```
pylint 4.0.3
astroid 4.0.2
Python 3.12.9 (main, Mar 18 2025, 03:28:39) [GCC 12.2.0]
```

This code gives no errors:

```python
from dataclasses import dataclass
from typing import TypeVar, Generic

T = TypeVar('T')

@dataclass
class Base(Generic[T]):
    a: T

@dataclass
class Sub(Base[str]):
    pass

Sub(a="harf")
```

But if you change to the Python 3.12 Generic syntax:

```python
from dataclasses import dataclass

@dataclass
class Base[T]:
    a: T

@dataclass
class Sub(Base[str]):
    pass

Sub(a="harf")
```

Then you get errors:

```
/b.py:11:0: E1123: Unexpected keyword argument 'a' in constructor call (unexpected-keyword-arg)
```
